hare/broker.go

```diff
@@ -326,124 +290,40 @@ func (b *Broker) HandleEligibility(ctx context.Context, em *types.HareEligibilit
 	return false
 }
 
-func (b *Broker) getInbox(id types.LayerID) chan any {
-	b.mu.RLock()
-	defer b.mu.RUnlock()
-	out, exist := b.outbox[id.Uint32()]
-	if !exist {
-		return nil
-	}
-	return out
-}
-
-func (b *Broker) handleEarlyMessage(logger log.Log, layer types.LayerID, nodeID types.NodeID, msg any) error {
-	b.mu.Lock()
-	defer b.mu.Unlock()
-	layerNum := layer.Uint32()
-	if _, exist := b.pending[layerNum]; !exist { // create buffer if first msg
-		b.pending[layerNum] = make([]any, 0, inboxCapacity)
-	}
-
-	// we want to write all buffered messages to a chan with InboxCapacity len
-	// hence, we limit the buffer for pending messages
-	if len(b.pending[layerNum]) == inboxCapacity {
-		logger.With().Warning("too many pending messages, ignoring message",
-			log.Int("inbox_capacity", inboxCapacity),
-			log.Stringer("smesher", nodeID))
-		return nil
-	}
-	b.pending[layerNum] = append(b.pending[layerNum], msg)
-	return nil
-}
-
-func (b *Broker) cleanOldLayers() {
-	b.mu.Lock()
-	defer b.mu.Unlock()
-
-	for i := b.minDeleted.Add(1); i.Before(b.latestLayer); i = i.Add(1) {
-		_, exist := b.outbox[i.Uint32()]
-
-		if !exist { // unregistered
-			b.minDeleted = b.minDeleted.Add(1)
-		} else { // encountered first still running layer
-			break
-		}
-	}
-}
-
 // Register a layer to receive messages
 // Note: the registering instance is assumed to be started and accepting messages.
-func (b *Broker) Register(ctx context.Context, id types.LayerID) (chan any, error) {
-	b.setLatestLayer(ctx, id)
-
-	// check to see if the node is still synced
-	if !b.Synced(ctx, id) {
-		return nil, errInstanceNotSynced
-	}
-	return b.createNewInbox(id), nil
-}
-
-func (b *Broker) createNewInbox(id types.LayerID) chan any {
+func (b *Broker) Register(ctx context.Context, id types.LayerID) (chan any,
+	error,
+) {
 	b.mu.Lock()
 	defer b.mu.Unlock()
 
-	if len(b.outbox) >= b.limit {
-		// unregister the earliest layer to make space for the new layer
-		// cannot call unregister here because unregister blocks and this would cause a deadlock
-		instance := b.minDeleted.Add(1)
-		delete(b.outbox, instance.Uint32())
-		b.minDeleted = instance
-		b.With().Info("unregistered layer due to maximum concurrent processes", instance)
+	if !id.After(b.latestLayer) { // should expect to update only newer layers, its a coding error if this happens
+		panic(fmt.Sprintf("broker register called with old layer, current layer: %v, given layer: %v", b.latestLayer, id))
```

Comment:
use `b.logger.With().Fatal()` to print the fields and avoid unnecessary stacktrace.

Reply:
The stack trace is useful to understand the context of the failure, it lets you know at least what component called this code, that's why I didn't use log.Fatal and why I would suggest not using log.Fatal except in cases where the stack is very obvious, e.g. parsing commandline args.

Reply:
i disagree. there is only one caller of Register. i can tell exactly the stack trace in this case. 
stack trace just filled up the log volume without extra information.

Reply:
I count 26 references to `Register` in our tests and if this were to happen in a test just calling log.Fatal might leave you wondering where the problem occurred, with a stack trace you'd know the line in the test that caused the problem.

Given that the node will be shutting down at this point, the addition of a few more lines to the log seems like an insignificant addition, not something to worry about.

Maybe I'm misunderstanding something about how our logging works?

Reply:
> Given that the node will be shutting down at this point, the addition of a few more lines to the log seems like an insignificant addition, not something to worry about.

by that same argument, i'd say reducing the log is also not something to worry about.

tests are not really my consideration. 
my concerns are about log volumes in grafana when we actually debug. or from the user submitted log.
the smaller the better. we are doing a concerted effort to reduce log volume and this does not help.